File: src/main/java/pyspark_elastic/PythonHelper.java
package pyspark_elastic;

import java.util.HashMap;
import java.util.Map;

import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.elasticsearch.hadoop.cfg.ConfigurationOptions;
import org.elasticsearch.spark.rdd.api.java.JavaEsSpark;

import pyspark_elastic.pickling.BatchPickle;

@SuppressWarnings({ "rawtypes", "unchecked" })
public class PythonHelper {
	public JavaRDD esRDD(JavaSparkContext sc, String resource, String query, Map<String, String> cfg) {
		cfg = readConfig(resource, query, cfg);
		return JavaEsSpark.esRDD(sc, cfg).mapPartitions(new BatchPickle(), true);
	}

	public JavaRDD esJsonRDD(JavaSparkContext sc, String resource, String query, Map<String, String> cfg) {
		cfg = readConfig(resource, query, cfg);
		return JavaEsSpark.esJsonRDD(sc, cfg).mapPartitions(new BatchPickle(), true);
	}
	
	public void saveJsonToEs(JavaRDD<String> rdd, String resource, Map<String, String> cfg) {
		cfg = writeConfig(resource, cfg);
		JavaEsSpark.saveJsonToEs(rdd, cfg);
	}

	private Map<String, String> readConfig(String resource, String query, Map<String, String> cfg) {
		if (cfg == null) {
			cfg = new HashMap<>();
		}
	
		if (resource != null) {
			cfg.put(ConfigurationOptions.ES_RESOURCE_READ, resource);
		}
	
		if (query != null) {
			cfg.put(ConfigurationOptions.ES_QUERY, query);
		}
		
		return cfg;
	}

	private Map<String, String> writeConfig(String resource, Map<String, String> cfg) {
		if (cfg == null) {
			cfg = new HashMap<>();
		}
	
		if (resource != null) {
			cfg.put(ConfigurationOptions.ES_RESOURCE_WRITE, resource);
		}
		
		return cfg;
	}
}

File: src/main/scala/pyspark_elastic/PythonHelper.scala
package pyspark_elastic

import java.util.{ Map => JMap }
import scala.collection.JavaConversions._
import scala.collection.mutable
import org.apache.spark.api.java._
import org.elasticsearch.spark._
import org.elasticsearch.spark.rdd.api.java.JavaEsSpark
import pyspark_util.Conversions._
import pyspark_util.Pickling._
import java.util.HashMap
import org.elasticsearch.hadoop.cfg.ConfigurationOptions.{ ES_RESOURCE_WRITE, ES_RESOURCE_READ, ES_QUERY }
import pyspark_util.Pickling
import org.apache.spark.rdd.RDD

class PythonHelper() {

  Pickling.register()

  def esJsonRDD(sc: JavaSparkContext, cfg: JMap[String, String]) = {
    JavaRDD.fromRDD(sc.sc.esJsonRDD(config(cfg)).pickle())
  }

  def saveJsonToEs(rdd: JavaRDD[Array[Byte]], cfg: JMap[String, String]) = {
    rdd.rdd.unpickle().asInstanceOf[RDD[String]].saveJsonToEs(config(cfg))
  }

  private[this] def config(cfg: JMap[String, String]) = {
    if (cfg != null) {
      mapAsScalaMap(cfg)
    } else {
      Map[String, String]()
    }
  }
}
